zap/src/main/java/org/zaproxy/zap/extension/history/SiteNotesAddDialog.java

```diff
@@ -0,0 +1,328 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2024 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.history;
+
+import java.awt.Component;
+import java.awt.Frame;
+import java.awt.HeadlessException;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.Vector;
+import javax.swing.Box;
+import javax.swing.DefaultListModel;
+import javax.swing.GroupLayout;
+import javax.swing.JButton;
+import javax.swing.JLabel;
+import javax.swing.JList;
+import javax.swing.JPanel;
+import javax.swing.JScrollPane;
+import javax.swing.ListCellRenderer;
+import javax.swing.ScrollPaneConstants;
+import javax.swing.border.EmptyBorder;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.db.DatabaseException;
+import org.parosproxy.paros.extension.AbstractDialog;
+import org.parosproxy.paros.model.HistoryReference;
+import org.parosproxy.paros.model.SiteNode;
+import org.parosproxy.paros.network.HttpMalformedHeaderException;
+import org.zaproxy.zap.utils.ZapTextArea;
+
+public class SiteNotesAddDialog extends AbstractDialog {
```

Comment:
IMO we should not need to constantly reopen the dialogue to edit the notes, it should be possible to edit them without the need to close the dialogue.

I don't think it should show the history references of child nodes, this gives the impression the parent nodes have more history references (and notes) than they actually have.

Reply:
> IMO we should not need to constantly reopen the dialogue to edit the notes, it should be possible to edit them without the need to close the dialogue.
> 
> I don't think it should show the history references of child nodes, this gives the impression the parent nodes have more history references (and notes) than they actually have.

for the first part are you suggesting to not extend SiteNote...from AbstractDialog (Dialog box) or make it appear as a proper window like requester window ?
and for the second part are you suggesting to only include the current Site History reference or pastHistoryReference of cur SiteNode and don't include the child nodes .

Reply:
I'm saying it should behave differently (it should still be a dialogue). e.g. if the user modifies a note and changes to another history reference it should either: apply the changes, prompt the user to apply them, or have a button to apply the changes.

Yes, only show the history references of the chosen site node.

Reply:
ok i got it, i think it should autosave the note maybe, if the user changes to another history reference.

Reply:
If the changes are automatically applied (I'm fine with that) the cancel/save buttons should be changed to close.

Reply:
> IMO we should not need to constantly reopen the dialogue to edit the notes, it should be possible to edit them without the need to close the dialogue.
> 
> I don't think it should show the history references of child nodes, this gives the impression the parent nodes have more history references (and notes) than they actually have.

actually in the second part of question,  most of site node has temporary Href in historyReference so the whole point of adding note feature to siteNode seems irrelevant , so i thought showing href of child nodes would be good as it will help user to not go deep of nodes to set or get the notes of particular href.

Reply:
In that case to which node is the note being added, the user will not (easily) know solely based on the ID.

Reply:
The new dialogue is to handle the case of multiple history references on the same node.

Reply:
ok got it, and sorry to bother you so much one last thing about the suggestion to put note options to only node which have atleasty one historyReference as permanent type as most of the parent node have temporary Href type and there is no use of putting note options for those node.

Reply:
Does not bother me. Yeah, that sounds good, the menu item can be shown as disabled for those cases.

Reply:
> If the changes are automatically applied (I'm fine with that) the cancel/save buttons should be changed to close.

I don't think we should remove save and cancel button but just implement something like if the user forget to press save button before changing the Href , it will save automatically( just for mistakes ) .

Reply:
So we save automatically then the user cancels, what's the expectation? The save button also closes the dialogue and we are back to square one that we need to reopen the dialogue again.

Reply:
Save button will not close the dialogue only cancel button will, and user have to press save button to save notes but in case he forgets and change the href (select another href ) then only it will get save automatically.

Reply:
If we provide the save button we should not save automatically, we can prompt the user instead. The cancel implies undoing changes, if it doesn't it should be called close.

Reply:
Sounds good.